java-frontend/src/main/java/org/sonar/java/se/ProgramState.java

```diff
@@ -50,21 +50,64 @@
 import java.util.Objects;
 import java.util.Set;
 import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
 
 public class ProgramState {
 
   public static class Pop {
 
     public final ProgramState state;
     public final List<SymbolicValue> values;
+    public final List<SymbolicValueSymbol> valuesAndSymbols;
 
-    public Pop(ProgramState programState, List<SymbolicValue> result) {
+    public Pop(ProgramState programState, List<SymbolicValueSymbol> result) {
       state = programState;
-      values = result;
+      values = result.stream().map(SymbolicValueSymbol::symbolicValue).collect(Collectors.toList());
+      valuesAndSymbols = result;
     }
 
   }
 
+  public static class SymbolicValueSymbol {
```

Comment:
This is, IMO worth a comment to explain why we have this class and why its equals/hashcode is implemented like this. And especially mention the fact that symbol is getting tracked here for sake of reporting.